tests/system/test_gbq.py

```diff
@@ -541,18 +588,13 @@ def test_zero_rows(self, project_id):
             credentials=self.credentials,
             dialect="legacy",
         )
-        page_array = np.zeros(
-            (0,),
-            dtype=[
-                ("title", object),
-                ("id", np.dtype(int)),
-                ("is_bot", np.dtype(bool)),
-                ("ts", "M8[ns]"),
-            ],
-        )
-        expected_result = DataFrame(
-            page_array, columns=["title", "id", "is_bot", "ts"]
-        )
+        empty_columns = {
+            "title": pandas.Series([], dtype=object),
+            "id": pandas.Series([], dtype=object),
+            "is_bot": pandas.Series([], dtype=object),
```

Comment:
But isn't this a worse result that previous - that an empty column reverts to `object` rather than its original type?

Reply:
Yes, it is worse. We'd have to add a second type-conversion step for any non-null object columns. I believe this case only happens on empty dataframes, so I guess it wouldn't be the worst to check for that.

Reply:
I am definitely keen on doing this well once, rather than us & `google-cloud-bigquery` repeating logic.

But here, it seems like `pandas-gbq` logic is _better_ - to read the schema, and then use that schema in the dataframe. No?

Reply:
Done in https://github.com/pydata/pandas-gbq/pull/247/commits/0a6e8d91c3417977880525ea88bc17feba5ebcee

Reply:
Great! 

Should we do this on every dataframe; no need to have a special case for empty ones?

Reply:
> Should we do this on every dataframe; no need to have a special case for empty ones?

We could (if we tell [astype()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.astype.html#pandas.Series.astype) to ignore errors, but I think it's unnecessary. The dataframe constructor seems to pick the right type when bools and ints are passed in: object if it contains nulls, bool/int64 otherwise.

I'm very tempted to use the new nullable integer type by default at https://github.com/pydata/pandas-gbq/issues/242, but probably want to wait for pandas to make that change. For the cases where folks want a very particular dtype, I propose in https://github.com/pydata/pandas-gbq/issues/242 that we add a dtypes argument, much as we did in google-cloud-bigquery.

Reply:
OK great, thanks @tswast